src/mixins/java/org/spongepowered/common/mixin/core/entity/player/ServerPlayerEntityMixin.java

```diff
@@ -468,26 +495,144 @@ public void teleportTo(net.minecraft.world.server.ServerWorld world, double x, d
      */
-    @javax.annotation.Nullable
-    @Overwrite
-    public Entity changeDimension(ServerWorld destination) {
+    @SuppressWarnings("ConstantConditions")
+    @Nullable
+    public Entity bridge$changeDimension(final ServerWorld originalDestinationWorld, final SpongeTeleporter spongeTeleporter) {
+        // Sponge Start
         if (this.shadow$getCommandSenderWorld().isClientSide || this.removed) {
             return (ServerPlayerEntity) (Object) this;
         }
 
-        final WrappedITeleporterPortalType portalType = new WrappedITeleporterPortalType((PlatformITeleporterBridge) destination.getPortalForcer(), null);
-
         try (final CauseStackManager.StackFrame frame = PhaseTracker.getCauseStackManager().pushCauseFrame()) {
             frame.pushCause(this);
-            frame.pushCause(portalType);
-            frame.addContext(EventContextKeys.MOVEMENT_TYPE, MovementTypes.PORTAL);
+            // frame.pushCause(spongeTeleporter); // TODO: Expose the SpongeTeleporter interface as Teleporter...ish?
+            frame.addContext(EventContextKeys.MOVEMENT_TYPE, spongeTeleporter.getMovementType());
 
-            EntityUtil.invokePortalTo((ServerPlayerEntity) (Object) this, portalType, destination);
-            return (ServerPlayerEntity) (Object) this;
-        }
+            final ChangeEntityWorldEvent.Pre preChangeEvent =
+                PlatformHooks.getInstance().getEventHooks().callChangeEntityWorldEventPre((Entity) (Object) this, originalDestinationWorld);
+            if (preChangeEvent.isCancelled()) {
+                return (Entity) (Object) this;
+            }
+            final ServerWorld targetWorld = (ServerWorld) preChangeEvent.getDestinationWorld();
+            final Vector3d currentPosition = VecHelper.toVector3d(this.position());
+            // Sponge End
+            this.isChangingDimension = true;
+            ServerWorld serverworld = this.shadow$getLevel();
+            RegistryKey<World> registrykey = serverworld.dimension();
+            if (registrykey == World.END && targetWorld.dimension() == World.OVERWORLD && spongeTeleporter.isVanilla()) { // avoids modded dimensions
+                this.shadow$unRide();
+                this.shadow$getLevel().removePlayerImmediately((ServerPlayerEntity) (Object) this);
+                if (!this.wonGame) {
+                    this.wonGame = true;
+                    this.connection.send(new SChangeGameStatePacket(SChangeGameStatePacket.WIN_GAME, this.seenCredits ? 0.0F : 1.0F));
+                    this.seenCredits = true;
+                }
+
+                // TODO: Figure out the reposition step (though that might just be a respawn)
+                return (Entity) (Object) this;
+            } else {
+                IWorldInfo iworldinfo = targetWorld.getLevelData();
+                this.connection.send(new SRespawnPacket(targetWorld.dimensionType(), targetWorld.dimension(),
```

Comment:
Once the server can run, I can figure out if we'll need to send fake respawn packets or not (to clear Vanilla client chunk caches).

May need to keep in mind, if not, will need to call the packet hook for this.